### ISSUE TITLE
`dg_ga_jvp_column_sum_v2` can break

### ISSUE THREAD

**Issue by patrick-kidger** [google-research/torchsde#46]
It assumes that the batch dimension can be modified, but something like e.g.
```
z = (batch, channel)
def g(self, t, y):
    return torch.stack([z, y], dim=2)
```
will then throw an error because of the different batch dimensions. (This being a scenario that I've actually run into.)

Incidentally I'm also concerned (but have not tested) that multiplying the batch dimension will put a peak in our memory usage. Having a dramatically non-rectangular memory profile could reduce the maximum possible batch size and thus slow things down overall.

I'm wondering about switching to `v1` for now? At least on `test_strat.py` I don't find it that much slower. (And as per above, may even be faster across a batch.)

**Comment by lxuechen** [comment]
The example seems interesting. I will make `v1` a default for now. If you could provide some more detail on how `z` is passed in, then I might be able to improve the usage of `v2`. Seems related, I remember running into something similar where I had to "contextualize" the SDE based on a representation produced by GRUs back doing latent SDEs. 

> Incidentally I'm also concerned (but have not tested) that multiplying the batch dimension will put a peak in our memory usage.

This could be true when the Brownian motion dimension is large. Though, if we use adjoints, then the issue might not be as prominent if we could fit models without this term with backprop through solver.

**Comment by patrick-kidger** [comment]
You're thinking that `additional_ys` represents this additional static state, and whilst we're at it we could add `additional_params` to augment `SDE.parameters()` for the adjoint?

If so I'd note that `additional_params` would only be needed in the adjoint case. We could follow `torchdiffeq` for consistency on this - there we called it `adjoint_params`, and if passed then it is used instead of the parameters of the vector field, rather than as well.

On the format of `additional_ys`: I'm quite keen to avoid explicitly encoding a single batch dimension.
I'd suggest essentially following what `autograd.Function` does on this: accept a tuple of Python objects; and if they're gradient-requiring tensors then compute gradients wrt them. Allow tensors to be of any shape.
This does mean that we can't really use `v2`, as we don't expect to have access to a batch dimension, but I think this kind of batch dimension hacking is quite fragile to the variety of things a user can throw at it anyway.

For speeding up `v1`, there is this: https://github.com/pytorch/pytorch/issues/42368 which mentions the possibility of a `torch.vmap`, in particular with a view to batch-vjps. I don't know the state of it but it might be interesting to us.

**Comment by patrick-kidger** [comment]
Actually thinking about - with the above proposal we wouldn't need an `adjoint_params`. Whatever extra tensors that we need to compute gradients with can just be included in `additional_ys` and ignored in the drift/diffusion.

**Comment by lxuechen** [comment]
Taking a step back, I think having `sdeint` take in `additional_ys` is likely going to overcomplicate the solver code. I'm not too inclined to do this at the moment. 

I do feel a need to support back-propagating gradients backward towards non-parameters nodes with adjoints. I am fully aware of `adjoint_params ` of torchdiffeq, and I can send in a PR on this.

**Comment by lxuechen** [comment]
Re: `torch.vmap` 

I'm not entirely sure this will make our lives easier. Given that there's not much documentation on what's going on there, much of this discussion seems rather like speculation in my opinion.
